chain/client/src/chunk_validation.rs

```diff
@@ -486,24 +486,107 @@ impl Client {
         Ok((main_transition, implicit_transitions, receipts_hash))
     }
 
+    /// State witness proves the execution of receipts proposed by `prev_chunk`.
+    /// This function collects all incoming receipts for `prev_chunk`, along with the proofs
+    /// that those receipts really originate from the right chunks.
+    /// TODO(resharding): `get_incoming_receipts_for_shard` generates invalid proofs on resharding
+    /// boundaries, because it removes the receipts that target the other half of a split shard,
+    /// which makes the proof invalid. We need to collect the original proof and later, after verifcation,
+    /// filter it to remove the receipts that were meant for the other half of the split shard.
+    fn collect_source_receipt_proofs(
+        &self,
+        prev_block: &Block,
+        prev_chunk_header: &ShardChunkHeader,
+    ) -> Result<HashMap<ChunkHash, ReceiptProof>, Error> {
+        // Find the first block that included `prev_chunk`.
+        // Incoming receipts were generated by the blocks before this one.
+        let prev_chunk_original_block: Cow<BlockHeader> = {
```

Comment:
why do we need `Cow` here?

Reply:
`prev_block.header()` returns `&BlockHeader`, but `self.chain.chain_store().get_block_header()` returns `BlockHeader`.I wanted to have a variable that can hold both (either owned or borrowed), so `Cow` seems like a good fit.
I could clone the header, but that feels wasteful. Especially since in the most common case (no missing chunks) the loop stops after the first iteration, so we can avoid clones altogether.

Reply:
just clone block header and don't worry about it :)
I suggest using advanced constructs such as `Cow` to improve performance only when it actually matters. Here it is clearly not the case, so please prefer a solution which is simpler.

Reply:
Okay, removed the Cow to keep things simple.